quesma/quesma/search_alternative.go

```diff
@@ -0,0 +1,263 @@
+			return nil, err
+		} else {
+			responseBody, err := io.ReadAll(resp.Body)
+
+			if err != nil {
+				return nil, err
+			}
+
+			if err := resp.Body.Close(); err != nil {
+				return nil, err
+			}
+
+			if resp.StatusCode != http.StatusOK {
+				return nil, fmt.Errorf("error calling elastic. got error code: %d", resp.StatusCode)
+			}
+			return responseBody, nil
+		}
+	}
+}
+
+func (q *QueryRunner) maybeCreatePancakeExecutionPlan(ctx context.Context, resolvedTableName string, plan *model.ExecutionPlan, queryTranslator IQueryTranslator, body types.JSON, table *clickhouse.Table, isAsync bool) (*model.ExecutionPlan, executionPlanExecutor) {
+
+	hasAggQuery := false
+	queriesWithoutAggr := make([]*model.Query, 0)
+	for _, query := range plan.Queries {
+		switch query.Type.AggregationType() {
+		case model.MetricsAggregation, model.BucketAggregation, model.PipelineAggregation:
+			hasAggQuery = true
+		default:
+			queriesWithoutAggr = append(queriesWithoutAggr, query)
+		}
+	}
+
+	if hasAggQuery {
+		if chQueryTranslator, ok := queryTranslator.(*queryparser.ClickhouseQueryTranslator); ok {
+
+			// TODO FIXME check if the original plan has count query
+			addCount := false
+
+			if pancakeQueries, err := chQueryTranslator.PancakeParseAggregationJson(body, addCount); err == nil {
+				logger.InfoWithCtx(ctx).Msgf("Running alternative pancake queries")
+				queries := append(queriesWithoutAggr, pancakeQueries...)
+				alternativePlan := &model.ExecutionPlan{
+					IndexPattern:          plan.IndexPattern,
+					QueryRowsTransformers: make([]model.QueryRowsTransformer, len(queries)),
+					Queries:               queries,
+					StartTime:             plan.StartTime,
+					Name:                  "pancake",
+				}
+
+				return alternativePlan, func(ctx context.Context) ([]byte, error) {
+
+					return q.executeAlternativePlan(ctx, plan, queryTranslator, table, body, false)
+				}
+
+			} else {
+				// TODO: change to info
+				logger.ErrorWithCtx(ctx).Msgf("Error parsing pancake queries: %v", err)
+			}
+		} else {
+			logger.ErrorWithCtx(ctx).Msgf("Alternative plan is not supported for non-clickhouse query translators")
+		}
+	}
+
+	return nil, nil
+}
+
+func (q *QueryRunner) executeAlternativePlan(ctx context.Context, plan *model.ExecutionPlan, queryTranslator IQueryTranslator, table *clickhouse.Table, body types.JSON, isAsync bool) (responseBody []byte, err error) {
+
+	doneCh := make(chan AsyncSearchWithError, 1)
+
+	q.transformQueries(ctx, plan, table)
+
+	if resp, err := q.checkProperties(ctx, plan, table, queryTranslator); err != nil {
+		return resp, err
+	}
+
+	q.runExecutePlanAsync(ctx, plan, queryTranslator, table, doneCh, nil)
+
+	response := <-doneCh
+
+	if response.err == nil {
+		if isAsync {
+			asyncResponse := queryparser.SearchToAsyncSearchResponse(response.response, "__quesma_alternative_plan", false, 200)
+			responseBody, err = asyncResponse.Marshal()
+			if err != nil {
```

Comment:
Seems that these checks can be extracted to common place, line 249

Reply:
Do you mean these checks can be shared between the 'main' and the 'alternative' plan execution?

Reply:
I meant that these 3 lines of code can be move below IMO